pkg/kubelet/secret/secret_manager.go

```diff
@@ -0,0 +1,246 @@
+		}
+		return tmpItem
+	}()
+	if item != nil {
+		item.Lock()
+		defer item.Unlock()
+		item.refCount++
+		item.lastUpdateTime = unixEpoch
+	}
+}
+
+func (s *secretStore) Delete(namespace, name string) {
+	key := objectKey{namespace: namespace, name: name}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if item, ok := s.items[key]; ok {
+		item.refCount--
+		if item.refCount == 0 {
+			delete(s.items, key)
+		}
+	}
+}
+
+func (s *secretStore) Get(namespace, name string) (*v1.Secret, error) {
+	key := objectKey{namespace: namespace, name: name}
+
+	item := func() *secretStoreItem {
+		s.lock.Lock()
+		defer s.lock.Unlock()
+		return s.items[key]
+	}()
+	if item == nil {
+		return nil, fmt.Errorf("secret %v/%v not registered", namespace, name)
+	}
+	item.Lock()
+	defer item.Unlock()
+	if item.err == nil && time.Now().Before(item.lastUpdateTime.Add(s.refreshFrequency)) {
+		return item.secret, item.err
+	}
+
+	secret, err := s.kubeClient.Core().Secrets(namespace).Get(name, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return item.secret, item.err
+	}
+	if !isSecretNewer(secret, item.secret) {
+		return item.secret, item.err
+	}
+	item.secret = secret
+	item.err = err
+	item.lastUpdateTime = time.Now()
+
+	return secret, err
+}
+
+// cachingSecretManager keeps a cache of all secrets necessary for registered pods.
+// It implements the following logic:
+// - whenever a pod is created or updated, the current versions of all its secrets
+//   are grabbed from apiserver and stored in local cache
+// - every GetSecret call is served from local cache
+// - every X seconds we are refreshing the local cache by grabbing current version
+//   of all registered secrets from apiserver
+type cachingSecretManager struct {
+	secretStore *secretStore
+
+	lock           sync.Mutex
+	registeredPods map[objectKey]*v1.Pod
+}
+
+func NewCachingSecretManager(kubeClient clientset.Interface) (Manager, error) {
+	csm := &cachingSecretManager{
+		secretStore:    newSecretStore(kubeClient, time.Minute),
+		registeredPods: make(map[objectKey]*v1.Pod),
+	}
+	return csm, nil
+}
+
+func (c *cachingSecretManager) GetSecret(namespace, name string) (*v1.Secret, error) {
+	return c.secretStore.Get(namespace, name)
+}
+
+// TODO: Before we will use secretManager in other places (e.g. for secret volumes)
+// we should update this function to also get secrets from those places.
+func getSecretNames(pod *v1.Pod) sets.String {
+	result := sets.NewString()
+	for _, reference := range pod.Spec.ImagePullSecrets {
+		result.Insert(reference.Name)
+	}
+	for i := range pod.Spec.Containers {
+		for _, envVar := range pod.Spec.Containers[i].Env {
+			if envVar.ValueFrom != nil && envVar.ValueFrom.SecretKeyRef != nil {
+				result.Insert(envVar.ValueFrom.SecretKeyRef.Name)
+			}
+		}
+	}
+	return result
+}
+
+func (c *cachingSecretManager) RegisterPod(pod *v1.Pod) {
+	names := getSecretNames(pod)
+	c.lock.Lock()
```

Comment:
I agree with @timstclair's previous comment. If we use locks, might as well make this function thread-safe. If it's not thread-safe, there is no use of locks.